### PR TITLE
helm: add ca.crt to tls secrets

### PR DESCRIPTION
**Pull request by kaworu**
This PR aim to bring compatibility with [cert-manager](https://cert-manager.io/) for the Hubble certificates.

Technically, the requirement to support cert-manager generated TLS certificates is to add and use the CA certificate as `ca.crt` to our `kubernetes.io/tls` Secret instead of projecting it from the `hubble-ca-cert` ConfigMap. From there, it should be possible to disable certificate generation for Hubble with `--set hubble.tls.auto.enabled=false` and have cert-manager generate both `hubble-server-certs` and `hubble-relay-client-certs` for Hubble Server and Relay, respectively.

This PR update the clustermesh certificates in the same way for consistency's sake.

There are three mechanism we use to generate TLS certificates:
- [x] Through Helm directly with `--set hubble.tls.auto.method=helm` (the default)
  * See d1c3c2012aaba2c7cdfb9351fabb161f01f8e668 ("helm: add ca.crt in all the generated TLS Secrets") from this PR
- [x] Through `certgen` with `--set hubble.tls.auto.method=cronJob` (used by `quick-hubble-install.yaml`)
  * https://github.com/cilium/certgen/pull/15
  * 161fd8f1682a534541a93430b68876daeb0ad3f4 ("helm: bump certgen to v0.1.4")
- [x] Through `cilium install` using [the cilium-cli](https://github.com/cilium/cilium-cli)
  * https://github.com/cilium/cilium-cli/pull/145

### CODE REVIEW

**Comment by kaworu** [comment]
### Test on GKE
deploy Cilium, notice `hubble.tls.auto.enabled=false`:
```
% helm install cilium ./install/kubernetes/cilium \
  --namespace kube-system \
  --set nodeinit.enabled=true \
  --set nodeinit.reconfigureKubelet=true \
  --set nodeinit.removeCbrBridge=true \
  --set cni.binPath=/home/kubernetes/bin \
  --set gke.enabled=true \
  --set ipam.mode=kubernetes \
  --set nativeRoutingCIDR=${NATIVE_CIDR:?} \
  --set image.tag=v1.9.5 \
  --set operator.image.tag=v1.9.5 \
  --set hubble.relay.image.tag=v1.9.5 \
  --set hubble.enabled=true \
  --set hubble.relay.enabled=true \
  --set hubble.tls.auto.enabled=false
```

At that point, we can observe that Hubble is waiting on the certificates to show up:

```console
% kubectl logs -n kube-system ds/cilium | grep subsys=hubble
…
level=info msg="Waiting for Hubble server TLS certificate and key files to be created" subsys=hubble
```

and the Relay Pod is in `ContainerCreating` state waiting on the Secrets as well:
```console
% kubectl describe -n kube-system pod hubble-relay-787666d6fc-xgqvv
…
Events:
  Type     Reason       Age                  From               Message
  ----     ------       ----                 ----               -------
  Normal   Scheduled    3m11s                default-scheduler  Successfully assigned kube-system/hubble-relay-787666d6fc-xgqvv to gke-alex-gke-cluster-default-pool-05f50d4f-943q
  Warning  FailedMount  3m10s                kubelet            MountVolume.SetUp failed for volume "tls" : failed to sync secret cache: timed out waiting for the condition
  Warning  FailedMount  68s                  kubelet            Unable to attach or mount volumes: unmounted volumes=[tls], unattached volumes=[config tls hubble-relay-token-wkwm9 hubble-sock-dir]: timed out waiting for the condition
  Warning  FailedMount  63s (x8 over 3m10s)  kubelet            MountVolume.SetUp failed for volume "tls" : secret "hubble-relay-client-certs" not found
```

Now install cert-manager by [following the documentation](https://cert-manager.io/docs/installation/kubernetes/#installing-with-helm):
```
…
helm install \
  cert-manager jetstack/cert-manager \
  --namespace cert-manager \
  --version v1.2.0 \
  --create-namespace \
  --set installCRDs=true
```
Configure cert-manager with [a CA Issuer](https://cert-manager.io/docs/configuration/ca/), and create the certificates needed by Hubble and Relay:
```console
% cat <<EOF > hubble-cert-manager.yaml
---
apiVersion: v1
kind: Secret
metadata:
  name: hubble-ca-keypair
  namespace: kube-system
data:
  tls.crt: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUMrVENDQWVHZ0F3SUJBZ0lKQUtQR3dLRGwvNUhuTUEwR0NTcUdTSWIzRFFFQkN3VUFNQk14RVRBUEJnTlYKQkFNTUNHcHZjMmgyWVc1c01CNFhEVEU1TURneU1qRTJNRFUxT0ZvWERUSTVNRGd4T1RFMk1EVTFPRm93RXpFUgpNQThHQTFVRUF3d0lhbTl6YUhaaGJtd3dnZ0VpTUEwR0NTcUdTSWIzRFFFQkFRVUFBNElCRHdBd2dnRUtBb0lCCkFRQ3doU0IvcVc2L2tMYjJ6cHUrRUp2RDl3SEZhcStRQS8wSkgvTGxseW83ekFGeCtISHErQ09BYmsrQzhCNHQKL0hVRXNuczVSTDA5Q1orWDRqNnBiSkZkS2R1UHhYdTVaVllua3hZcFVEVTd5ZzdPU0tTWnpUbklaNzIzc01zMApSNmpZbi9Ecmo0eFhNSkVmSFVEcVllU1dsWnIzcWkxRUZhMGM3ZlZEeEgrNHh0WnROTkZPakg3YzZEL3ZXa0lnCldRVXhpd3Vzc2U2S01PV2pEbnYvNFZyamVsMlFnVVlVYkhDeWVaSG1jdGkrSzBMV0Nmby9SZzZQdWx3cmJEa2gKam1PZ1l0MzBwZGhYME9aa0F1a2xmVURIZnA4YmpiQ29JMnRhWUFCQTZBS2pLc08zNUxBRVU3OUNMMW1MVkh1WgpBQ0k1VWppamEzVlBXVkhTd21KUEp5dXhBZ01CQUFHalVEQk9NQjBHQTFVZERnUVdCQlFtbDVkVEFaaXhGS2hqCjkzd3VjUldoYW8vdFFqQWZCZ05WSFNNRUdEQVdnQlFtbDVkVEFaaXhGS2hqOTN3dWNSV2hhby90UWpBTUJnTlYKSFJNRUJUQURBUUgvTUEwR0NTcUdTSWIzRFFFQkN3VUFBNElCQVFCK2tsa1JOSlVLQkxYOHlZa3l1VTJSSGNCdgpHaG1tRGpKSXNPSkhac29ZWGRMbEcxcFpORmpqUGFPTDh2aDQ0Vmw5OFJoRVpCSHNMVDFLTWJwMXN1NkNxajByClVHMWtwUkJlZitJT01UNE1VN3ZSSUNpN1VPbFJMcDFXcDBGOGxhM2hQT2NSYjJ5T2ZGcVhYeVpXWGY0dDBCNDUKdEhpK1pDTkhCOUZ4alNSeWNiR1lWaytUS3B2aEphU1lOTUdKM2R4REthUDcrRHgzWGNLNnNBbklBa2h5SThhagpOVSttdzgvdG1Sa1A0SW4va1hBUitSaTBxVW1Iai92d3ZuazRLbTdaVXkxRllIOERNZVM1TmtzbisvdUhsUnhSClY3RG5uMDM5VFJtZ0tiQXFONzJnS05MbzVjWit5L1lxREFZSFlybjk4U1FUOUpEZ3RJL0svQVRwVzhkWAotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==
  tls.key: LS0tLS1CRUdJTiBSU0EgUFJJVkFURSBLRVktLS0tLQpNSUlFb3dJQkFBS0NBUUVBc0lVZ2Y2bHV2NUMyOXM2YnZoQ2J3L2NCeFdxdmtBUDlDUi95NVpjcU84d0JjZmh4CjZ2Z2pnRzVQZ3ZBZUxmeDFCTEo3T1VTOVBRbWZsK0krcVd5UlhTbmJqOFY3dVdWV0o1TVdLVkExTzhvT3praWsKbWMwNXlHZTl0N0RMTkVlbzJKL3c2NCtNVnpDUkh4MUE2bUhrbHBXYTk2b3RSQld0SE8zMVE4Ui91TWJXYlRUUgpUb3grM09nLzcxcENJRmtGTVlzTHJMSHVpakRsb3c1Ny8rRmE0M3Bka0lGR0ZHeHdzbm1SNW5MWXZpdEMxZ242ClAwWU9qN3BjSzJ3NUlZNWpvR0xkOUtYWVY5RG1aQUxwSlgxQXgzNmZHNDJ3cUNOcldtQUFRT2dDb3lyRHQrU3cKQkZPL1FpOVppMVI3bVFBaU9WSTRvMnQxVDFsUjBzSmlUeWNyc1FJREFRQUJBb0lCQUNFTkhET3JGdGg1a1RpUApJT3dxa2UvVVhSbUl5MHlNNHFFRndXWXBzcmUxa0FPMkFDWjl4YS96ZDZITnNlanNYMEM4NW9PbmtrTk9mUHBrClcxVS94Y3dLM1ZpRElwSnBIZ09VNzg1V2ZWRXZtU3dZdi9Fb1V3eHFHRVMvcnB5Z1drWU5WSC9XeGZGQlg3clMKc0dmeVltbXJvM09DQXEyLzNVVVFiUjcrT09md3kzSHdUdTBRdW5FSnBFbWU2RXdzdWIwZzhTTGp2cEpjSHZTbQpPQlNKSXJyL1RjcFRITjVPc1h1Vm5FTlVqV3BBUmRQT1NrRFZHbWtCbnkyaVZURElST3NGbmV1RUZ1NitXOWpqCmhlb1hNN2czbkE0NmlLenUzR0YwRWhLOFkzWjRmeE42NERkbWNBWnphaU1vMFJVaktWTFVqbVlQSEUxWWZVK3AKMkNYb3dNRUNnWUVBMTgyaU52UEkwVVlWaUh5blhKclNzd1YrcTlTRStvVi90U2ZSUUNGU2xsV0d3KzYyblRiVwpvNXpoL1RDQW9VTVNSbUFPZ0xKWU1LZUZ1SWdvTEoxN1pvWjN0U1czTlVtMmRpT0lPSHorcTQxQzM5MDRrUzM5CjkrYkFtVmtaSFA5VktLOEMraS9tek5mSkdHZEJadGIweWtTM2t3OUIxTHdnT3o3MDhFeXFSQ2tDZ1lFQTBXWlAKbzF2MThnV2tMK2FnUDFvOE13eDRPZlpTN3dKY3E0Z0xnUWhjYS9pSkttY0x0RFN4cUJHckJ4UVo0WTIyazlzdQpzTFVrNEJobGlVM29iUUJNaUdtMGtITHVBSEFRNmJvdWZBMUJwZjN2VFdHSkhSRjRMeFJsNzc2akw4UXI4VnpxClpURVBtY0R0T0hpYjdwb2I1Z2IzSDhiVGhYeUhmdGZxRW55alhFa0NnWUVBdk9DdDZZclZhTlQrWThjMmRFYk4Kd3dJOExBaUZtdjdkRjZFUjlCODJPWDRCeGR0WTJhRDFtNTNqN2NaVnpzNzFYOE1TN25FcDN1dkFqaElkbDI3KwpZbTJ1dUUyYVhIbDN5VTZ3RzBETFpUcnVIU0Z5TVI4ZithbHRTTXBDd0s1NXluSGpHVFp6dXpYaVBBbWpwRzdmCk1XbVRncE1IK3puc3UrNE9VNFBHUW9FQ2dZQWNqdUdKbS84YzlOd0JsR2lDZTJIK2JGTHhSTURteStHcm16QkcKZHNkMENqOWF3eGI3aXJ3MytjRGpoRUJMWExKcjA5YTRUdHdxbStrdElxenlRTG92V0l0QnNBcjVrRThlTVVBcAp0djBmRUZUVXJ0cXVWaldYNWlaSTNpMFBWS2ZSa1NSK2pJUmVLY3V3aWZKcVJpWkw1dU5KT0NxYzUvRHF3Yk93CnRjTHAwUUtCZ0VwdEw1SU10Sk5EQnBXbllmN0F5QVBhc0RWRE9aTEhNUGRpL2dvNitjSmdpUmtMYWt3eUpjV3IKU25QSG1TbFE0aEluNGMrNW1lbHBDWFdJaklLRCtjcTlxT2xmQmRtaWtYb2RVQ2pqWUJjNnVGQ1QrNWRkMWM4RwpiUkJQOUNtWk9GL0hOcHN0MEgxenhNd1crUHk5Q2VnR3hhZ0ZCekxzVW84N0xWR2h0VFFZCi0tLS0tRU5EIFJTQSBQUklWQVRFIEtFWS0tLS0tCg==
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: hubble-ca
  namespace: kube-system
spec:
  ca:
    secretName: hubble-ca-keypair
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: hubble-server-certs
  namespace: kube-system
spec:
  dnsNames:
    - '*.default.hubble-grpc.cilium.io'
  usages:
    - server auth
  issuerRef:
    name: hubble-ca
  secretName: hubble-server-certs
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: hubble-relay-client-certs
  namespace: kube-system
spec:
  dnsNames:
    - '*.hubble-relay.cilium.io'
  usages:
    - client auth
  issuerRef:
    name: hubble-ca
  secretName: hubble-relay-client-certs
EOF
% kubectl apply -f hubble-cert-manager.yaml
secret/hubble-ca-keypair created
issuer.cert-manager.io/hubble-ca created
certificate.cert-manager.io/hubble-server-certs created
certificate.cert-manager.io/hubble-relay-client-certs created
```
Check Hubble and Relay:
```console
% kubectl get secrets -n kube-system hubble-relay-client-certs hubble-server-certs
NAME                        TYPE                DATA   AGE
hubble-relay-client-certs   kubernetes.io/tls   3      2m15s
hubble-server-certs         kubernetes.io/tls   3      2m15s
% kubectl logs -n kube-system ds/cilium | grep subsys=hubble
Found 2 pods, using pod/cilium-d5qzv
level=info msg="Configuring Hubble server" eventQueueSize=4096 maxFlows=4095 subsys=hubble
level=info msg="Starting local Hubble server" address="unix:///var/run/cilium/hubble.sock" subsys=hubble
level=info msg="Waiting for Hubble server TLS certificate and key files to be created" subsys=hubble
level=info msg="Certificate authority updated" config=tls-server subsys=hubble
level=info msg="Keypair updated" config=tls-server keypair-sn=a38a35bc2042874fc3888e66fb46b784 subsys=hubble
level=info msg="Starting Hubble server" address=":4244" subsys=hubble
% kubectl get pods -n kube-system hubble-relay-787666d6fc-xgqvv
NAME                            READY   STATUS    RESTARTS   AGE
hubble-relay-787666d6fc-xgqvv   1/1     Running   0          12m
```

Confirm status with the Hubble CLI:

```console
% kubectl port-forward -n kube-system deployment/hubble-relay 4245:4245 &
% hubble status --server localhost:4245
Healthcheck (via localhost:4245): Ok
Current/Max Flows: 8192/8192 (100.00%)
Flows/s: 11.30
Connected Nodes: 2/2
```

File: install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml [hunk 1]
```diff
@@ -554,16 +554,12 @@ spec:
...
               items:
+                - key: ca.crt
+                  path: client-ca.crt
                 - key: tls.crt
                   path: server.crt
                 - key: tls.key
                   path: server.key
               optional: true
-          - configMap:
-              name: hubble-ca-cert
```
Comment (gandro) [inline install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:573]: Is there anything blocking us from removing the `hubble-ca-cert` ConfigMap completely?

**Review by gandro — COMMENTED** [review]
Very nice work! One question that came up during review

File: install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml (hunk 1 above)
Comment (kaworu) [reply install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:573]: Only my strong opinion that storing it in a `ConfigMap` is the technically correct way. I just think you should be able to grant access to the CA certificate without granting access to a keypair.

if we were to remove it, I would suggest to do it the next release (i.e. v1.11) once we can be sure no Pod needs it during upgrade. Does that make sense?

Comment (gandro) [reply install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:573]: Yes, that does make sense! If we're planning to deprecate the ConfigMap, it's probably best to add a comment in the Helm template that it is supposed to be removed.

Comment (kaworu) [reply install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:573]: As I'm not intending to remove it I don't intent to deprecated it either, but I'm open to change my mind.

Currently, I feel that not using it is not enough to consider removing it. I would err on the safe side of keeping it around for a release and see how things goes, eventually deprecating it later if we realize it doesn't make sense to have access to it through a `ConfigMap`.

Comment (gandro) [reply install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:573]: I'm not sure I follow then. What's the usecase of deploying a `ConfigMap` if it is not referenced by our deployment at all? If we need to revert, then yeah we can revert the change and reintroduce it.

But I have a hard time understanding what the long-term plan for the ConfigMap should be after 1.10 with this PR, if not removal in 1.11.

Comment (kaworu) [reply install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:573]: That's fair, we could mark it deprecated for v1.10 and see how that goes.

Comment (kaworu) [reply install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:573]: @gandro I've added a comment in the `hubble-ca-configmap.yaml` template hinting that it is deprecated and will be removed in v1.11, please take another a look.

**Review by gandro — APPROVED** [review]
:100:

**Comment by kaworu** [comment]
PR ready now that [certgen v1.0.4 has been released](https://github.com/cilium/certgen/releases/tag/v0.1.4).

File: install/kubernetes/Makefile [hunk 2]
```diff
@@ -8,7 +8,7 @@ HUBBLE_UI_VERSION := "v0.7.3"
 MANAGED_ETCD_VERSION := "v2.0.7"
 ETCD_VERSION := "v3.4.13"
 NODEINIT_VERSION := "62bfbe88c17778aad7bef9fa57ff9e2d4a9ba0d8"
-CERTGEN_VERSION := "v0.1.3"
+CERTGEN_VERSION := "v0.1.4"
```
Comment (christarazi) [inline install/kubernetes/Makefile:11]: The commit SHA referenced in the commit msg has gone stale I think

File: install/kubernetes/cilium/templates/clustermesh-apiserver-admin-secret.yaml [hunk 3]
```diff
@@ -1,15 +1,18 @@
 {{- if and .Values.agent (not .Values.preflight.enabled) (or .Values.externalWorkloads.enabled .Values.clustermesh.useAPIServer) }}
-{{- if or (and (.Values.clustermesh.apiserver.tls.auto.enabled) (eq .Values.clustermesh.apiserver.tls.auto.method "helm")) (and .Values.clustermesh.apiserver.tls.admin.cert .Values.clustermesh.apiserver.tls.admin.key) }}
+{{- $adminCertsProvided := and .Values.clustermesh.apiserver.tls.ca.cert .Values.clustermesh.apiserver.tls.admin.cert .Values.clustermesh.apiserver.tls.admin.key }}
```
Comment (christarazi) [inline install/kubernetes/cilium/templates/clustermesh-apiserver-admin-secret.yaml:2]: Same issue with the commit SHA

**Review by christarazi — APPROVED** [review]
LGTM, minor nits inside the commit msgs

**Comment by kaworu** [comment]
@christarazi thanks for the review and good catch for the commit SHA. I guess they change after each rebase so I just left them out assuming that the locality would be enough context (if that make sense).

**Comment by aditighag** [comment]
>ERROR:CUSTOM: Please avoid long commit subjects (max: 75, found: 83)

checkpatch warning can be ignored. All other tests have passed; merging the PR.

**Pull request merged**